### PR TITLE
M5-3-3: Excude binary operators.

### LINKED ISSUE #363: `M5-3-3`: triggers on bitwise & operator
### Affected rules
 - M5-3-3

### Description

M5-3-3 reports that "Unary `&` operator shall not be overloaded" on instances of the *binary* `&` operator

### Example

```cpp
constexpr byte operator&(const byte lhs, const byte rhs) noexcept {
  return to_byte(to_underlying(lhs) & to_underlying(rhs));
}
```

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Modify this rule to only apply to unary `operator&`.

Fixes #363.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `M5-3-3`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: change_notes/2024-01-14-m5-3-3-exclude-binary.md [hunk 1]
```diff
@@ -0,0 +1 @@
+ * `M5-3-3` - exclude binary user defined `operator&` from this rule.
```
Comment (rvermeulen) [inline change_notes/2024-01-14-m5-3-3-exclude-binary.md:1]: Apply format from developer handbook's style guide.

```suggestion
`M5-3-3` - `UnaryOperatorOverloaded.ql`:
  - exclude binary user defined `operator&` from this rule.
```

File: cpp/autosar/src/rules/M5-3-3/UnaryOperatorOverloaded.ql [hunk 2]
```diff
@@ -15,5 +15,12 @@ import cpp
...
 
 from Operator o
-where not isExcluded(o, OperatorsPackage::unaryOperatorOverloadedQuery()) and o.hasName("operator&")
+where
+  not isExcluded(o, OperatorsPackage::unaryOperatorOverloadedQuery()) and
+  o.hasName("operator&") and
+  (
+    if o instanceof MemberFunction
+    then o.getNumberOfParameters() = 0
+    else o.getNumberOfParameters() = 1
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/M5-3-3/UnaryOperatorOverloaded.ql:24]: Optionally capture this in the operator module.

**Review by rvermeulen — CHANGES_REQUESTED** [review]
Changenote style change request. Other than that looks good.

File: change_notes/2024-01-14-m5-3-3-exclude-binary.md (hunk 1 above)
Comment (lcartey) [reply change_notes/2024-01-14-m5-3-3-exclude-binary.md]: Done.

File: cpp/autosar/src/rules/M5-3-3/UnaryOperatorOverloaded.ql (hunk 2 above)
Comment (lcartey) [reply cpp/autosar/src/rules/M5-3-3/UnaryOperatorOverloaded.ql]: Done.

**Review by rvermeulen — APPROVED** [review]
Looks good.

**Pull request merged**